drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,614 @@
+	struct sdhc_host_props *props = &data->props;
+
+	memset(props, 0, sizeof(struct sdhc_host_props));
+
+	props->f_min = sdhc_config->min_freq;
+	props->f_max = sdhc_config->max_freq;
+	props->power_delay = sdhc_config->power_delay_ms;
+	props->host_caps.vol_330_support = true;
+	props->host_caps.vol_180_support = sdhc_config->support_1_8_v;
+	props->host_caps.vol_300_support = false;
+	props->host_caps.bus_8_bit_support = (sdhc_config->bus_width == 8);
+	props->host_caps.bus_4_bit_support = (sdhc_config->bus_width == 4);
+	props->host_caps.hs200_support = false;
+	props->host_caps.hs400_support = false;
+	props->host_caps.high_spd_support = true;
+	props->host_caps.sdr50_support = true;
+	props->host_caps.sdio_async_interrupt_support = true;
+	props->is_spi = false;
+}
+
+static int sdhc_stm32_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	struct sdhc_stm32_data *data = dev->data;
+
+	memcpy(props, &data->props, sizeof(struct sdhc_host_props));
+
+	return 0;
+}
+
+static int sdhc_stm32_get_card_present(const struct device *dev)
+{
+	int res;
+	uint8_t data = 0;
+	struct sdhc_stm32_data *dev_data = dev->data;
+
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&dev_data->bus_mutex, K_FOREVER);
+
+	HAL_SDIO_DirectCmd_TypeDef arg = {
+		.Reg_Addr = SDIO_CCCR_BUS_IF, .ReadAfterWrite = 0, .IOFunctionNbr = 0};
+
+	res = HAL_SDIO_ReadDirect(&dev_data->hsd, &arg, &data);
+
+	if (res != HAL_OK) {
+		sdhc_stm32_log_err_type(&dev_data->hsd);
+	}
+
+	k_mutex_unlock(&dev_data->bus_mutex);
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	/* Card is considered present if the read command did not time out */
+	return (res == HAL_OK);
+}
+
+static int sdhc_stm32_card_busy(const struct device *dev)
+{
+	struct sdhc_stm32_data *data = dev->data;
+
+	HAL_SDIO_StateTypeDef state = HAL_SDIO_GetState(&data->hsd);
+
+	return (state == HAL_SDIO_STATE_BUSY);
+}
+
+static int sdhc_stm32_reset(const struct device *dev)
+{
+	struct sdhc_stm32_data *data = dev->data;
+
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+	/* Reseting Host controller */
+	(void)SDMMC_PowerState_OFF(data->hsd.Instance);
+	k_msleep(data->props.power_delay);
+	(void)SDMMC_PowerState_ON(data->hsd.Instance);
+	k_msleep(data->props.power_delay);
+
+	/* Reseting card */
+	int res = HAL_SDIO_CardReset(&data->hsd);
+	if(res != HAL_OK) {
+		LOG_ERR("Card reset failed.\n");
+	}
+
+	k_mutex_unlock(&data->bus_mutex);
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	return res;
+}
+
+static const struct sdhc_driver_api sdhc_stm32_api = {
```

Comment:
Use the device API wrapper macro.

```suggestion
static DEVICE_API(sdhc, sdhc_stm32_api) = {
```